libsolc/libsolc.cpp

```diff
@@ -38,6 +40,28 @@ using namespace solidity;
 namespace
 {
 
+// The strings in this list must not be resized after they have been added here (via solidity_alloc()), because
+// this may potentially change the pointer that was passed to the caller from solidity_alloc().
+static list<string> solidityAllocations;
+
+/// Find the equivalent to @p _data in the list of allocations of solidity_alloc(),
+/// removes it from the list and returns its value.
+///
+/// If any invalid argument is being passed, it is considered a programming error
+/// on the caller-side and hence, will call abort() then.
+string takeOverAllocation(char const* _data)
+{
+	for (auto iter = begin(solidityAllocations); iter != end(solidityAllocations); ++iter)
+		if (iter->data() == _data)
+		{
+			string chunk = move(*iter);
+			solidityAllocations.erase(iter);
+			return chunk;
+		}
+
+	abort();
```

Comment:
Since this is actually called by us internally and not externally, we could actually handle this gracefully - so e.g. throw here and then catch in ``wrapReadCallback`` and return a ``ReadCallback::Result`` with ``success = false;`` and ``responseOrErrorMessage = "Callback returned memory not allocated by solidity_alloc";`` there or something like that...

Reply:
Ah, ok, you also use this in ``solidity_free`` as well - but actually that should be fine.

Reply:
> so e.g. throw here and then catch in wrapReadCallback

Thought about that, but I think this should be more serious. Then we get into the grey area that should `_reset` and `_free` also be lenient?

If consumers of the API can't get this single thing right, then what else should we expect. Note, most C APIs behave like this, they don't try to get you saved from all these issues, but usually they just leave this "UB", e.g. don't abort early.

Reply:
Fair enough!